### PR TITLE
Title: Decoding & Audio Transfers

### PR DESCRIPTION
**Pull request by styagi130**
Title: Decoding & Audio Transfers

**Source branch:** `decoding-audio-transfers`  
**Target branch:** `main`  
**Head:** `2a47b428b41827ed4a32529eb5e82936b1e6e609`  
**Comparison base:** local `main` at `ea1ebf55b25fceff372f885b4e7b5803f3d33dea`.

# What does this PR do ?

Reduce repeated work and host/device transfers in the EasyMagpie generation path. Keep feedback on the GPU, batch compatible codec work and waveform copies, and return only the valid audio from each terminal chunk.

**Collection**: TTS (`easymagpie_vllm_omni`).

## Scope against main

This is MR 2 of 5. Its branch includes the following prerequisite changes in the current diff to `main`:

- **Serving & Benchmarks** (`serving-benchmarks`).

The changelog below describes the additional work in this layer. The full current comparison is **9 commits; 45 files changed, 4331 insertions(+), 397 deletions(-)**. Merge the preceding MRs first. If their commits are preserved, those changes drop out of this diff as `main` advances. After a squash or rebase, restack and revalidate this branch before merging.

# Changelog

- Keep audio, phoneme and sticky end-of-sequence feedback in owning GPU storage, paired with asynchronous output and eager postprocessing.
- Batch compatible codec chunks, carry valid sample counts alongside terminal padding, and trim each request independently. Update the late-close regression for the padded codec contract.
- Replace repeated finite scalar dequantization arithmetic with a lookup table and fuse the HalfSnake residual addition with its activation kernel.
- Evaluate only the causal local-transformer prefix needed for each codebook prediction while preserving masking, temperature and sampling behavior.
- Pack eligible waveforms into a single CPU transfer and reconstruct their shapes and request order; retain upstream fallback for unsupported payloads.
- Select the codec generation worker in the ordinary deployment YAML, document its activation, and add numerical, state, ownership and CUDA graph regressions.

# Usage

Use the ordinary `tools/easymagpie_vllm_omni/deploy/easymagpie.yaml` profile. It selects `EasyMagpieCodecGPUGenerationWorker` for packed waveform transfers. Custom profiles must select that worker to use this path. Existing HTTP/WebSocket clients and voice inputs continue to use the same interface.

# GitHub Actions CI

Remote CI is pending. Trusted PRs use the existing copy-pr-bot flow. For an untrusted PR, a maintainer can trigger CI with `/ok to test <head-sha>` after publication.

## Validation

Recorded at `2a47b428b41827ed4a32529eb5e82936b1e6e609` on **2026-09-15**: **477 tests passed, zero failures and zero skips** in the complete CPU/CUDA serving suite. Tests used physical **H100 device 1**, exposed as the sole logical CUDA device in the container.

The package was freshly installed from this head. Source hashes, installed module origins, `pip check`, dependency invariance and author-matching DCO sign-offs passed. Full-branch pre-commit passed from the recorded main base. The runtime used vLLM/vLLM-Omni 0.26.0, PyTorch 2.11.0+cu130 and Triton 3.6.0.

<details>
<summary>Recorded validation commands</summary>

The serving command ran inside the pinned test container after installing this branch and providing the codec contract fixture:

```bash
/usr/bin/python3 -I -B /recipe/installed_pytest.py \
  --rootdir=/snapshot \
  --confcutdir=/snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  /snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  -p no:cacheprovider -q -ra --junitxml=/results/tests.xml \
  --durations=10
```

Full-branch formatting checks and the current diff check:

```bash
python3 -m pre_commit run \
  --from-ref ea1ebf55b25fceff372f885b4e7b5803f3d33dea \
  --to-ref 2a47b428b41827ed4a32529eb5e82936b1e6e609
git diff --check ea1ebf55b25fceff372f885b4e7b5803f3d33dea 2a47b428b41827ed4a32529eb5e82936b1e6e609
```

</details>

Live HTTP/WebSocket audio serving, cancellation/recovery, shutdown/drain, full-utterance quality/completion, performance acceptance and remote CI remain pending. The local test counts do not establish benchmark throughput or latency.

# Before your PR is "Ready for review"

- [x] Add regression tests and documentation with the implementation.
- [x] Pass the complete installed CPU/CUDA serving suite at the recorded head.
- [x] Pass full-branch pre-commit, diff checks and author-matching DCO checks.
- [ ] Complete contributor-guideline review and reviewer assessment of tests, documentation and optional dependency guards.
- [ ] Complete the remaining live serving, quality/completion and performance checks; attach portable validation evidence and pass remote CI.

**PR Type**: New Feature.

## Who can review?

TTS serving and codec/kernel maintainers. The six focused commits in this layer can be reviewed individually.

# Additional Information

Related to NVIDIA-NeMo/Speech#16220. Merge order: `serving-benchmarks` → `decoding-audio-transfers` → `voice-preloading` → `input-preparation-routing` → `h100-deployment`.

### CODE REVIEW

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py [hunk 1]
```diff
@@ -1109,8 +1109,8 @@ def postprocess(self, hidden_states: torch.Tensor, multimodal_outputs: Optional[
         out: dict[str, Any] = {}
         mm = multimodal_outputs or {}
         # The codes key depends on the emission mode (see make_omni_output):
-        # single-stage uses "model_outputs", two-stage uses "audio_codes" /
-        # nested "codes.audio". Read whichever is present.
+        # single-stage uses "model_outputs", two-stage uses "codes.audio".
+        # Keep accepting the legacy "audio_codes" key from older callers.
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py]: do we need to support legacy? codes are passed to codec via connector and to postporocess. both rely on nested one.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py [hunk 2]
```diff
@@ -81,6 +81,27 @@ def _build_omni_async_snapshot_payload(
         payload.setdefault("hidden_states", hidden_states[:, :0])
         return payload
 
+    def _update_intermediate_buffer(self, req_id: str, upd: dict) -> None:
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py:84]: was not clear why this is needed.
seems top-level keys (last_audio_codes, last_phoneme_token, phoneme_ended) are moved to CPU because _store_value treats them as non-resident. add a comment.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py (hunk 1 above)
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py:1113]: why do we care about accepting "audio_codes"? codes are passed to codec and to postprocess. both take nested structure. why do we need to maintain legacy?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py (hunk 2 above)
Comment (vklimkov-nvidia) [reply tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py:84]: upon checking upstream, nested structure is purposely encouraged. see other models declaration, for ex. https://github.com/vllm-project/vllm-omni/blob/v0.26.0/vllm_omni/model_executor/models/indextts2/indextts2_talker.py#L109-L116. 

i think we should do the same: `{"next_step": {"last_audio_codes": ..., "last_phoneme_token": ...}}`
not sure if `next_step` is the best name, other options - appreciated.

**Review by vklimkov-nvidia — COMMENTED** [review]
dd1cc1109be2511450a1f65cc76b7b83fa4b00b0 - mostly LGTM, small nit changes are requested.

i'd also want to see in the commit message what is the effect on throughput/latency of switch to async.

commit is small enough so it wasn't too problematic to review but it still contains unrelated changes which are not mentioned in the commit message:
* async output + owning GPU feedback
* device side `phoneme_ended`
* clean up: drop last_hidden and audio_codes.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/model.py [hunk 3]
```diff
@@ -120,9 +121,42 @@ def _payload_codes(
...
-        return torch.cat(packed, dim=0), frame_counts
+            empty = torch.empty((0, q), dtype=torch.long, device=device)
+            return empty, [], []
+        if all(rows.device.type == "cpu" for rows in packed):
+            codes = torch.cat(packed, dim=0).to(device=device, non_blocking=True)
+        else:
+            codes = torch.cat([rows.to(device=device, non_blocking=True) for rows in packed], dim=0)
+        return codes, frame_counts, valid_sample_counts
+
+    def _valid_sample_count(self, rows: torch.Tensor) -> int:
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/model.py:134]: i'd consider refactoring this. there is no need to copy entire codes to CPU and inspect them in order to find valid frames. you just need to attach metadata which describes valid frames. you already have this information before padding. 

the only difficulty to be aware of - is handling EOS during frame stacking. stage-0 stops decoding when EOS is detected. it can be either at the very last frame or at one but last frame. perhaps it is not even that important, we use frame_stacking <= 2, so at most we will lose 40ms of speech (1 frame at 25fps). perhaps just pass valid_frames as metadata and get rid of this redundant function

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/model.py [hunk 4]
```diff
@@ -120,9 +121,42 @@ def _payload_codes(
...
+            valid_sample_counts.append(self._valid_sample_count(rows))
         if not packed:
-            return torch.empty((0, q), dtype=torch.long, device=device), []
-        return torch.cat(packed, dim=0), frame_counts
+            empty = torch.empty((0, q), dtype=torch.long, device=device)
+            return empty, [], []
+        if all(rows.device.type == "cpu" for rows in packed):
+            codes = torch.cat(packed, dim=0).to(device=device, non_blocking=True)
+        else:
+            codes = torch.cat([rows.to(device=device, non_blocking=True) for rows in packed], dim=0)
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/model.py:131]: when would they be mixed though and when they would be not on CPU?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/packed.py [hunk 5]
```diff
@@ -422,52 +421,6 @@ def _one(self, inputs: torch.Tensor, page: torch.Tensor, has_initial: bool) -> t
         state.copy_(joined[-1:])
         return self.activation(outputs)
 
-    def _uniform_cuda(
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/packed.py:425]: why do we drop this? we keep uniform_cuda for PackedCausalConv1d but remove for transposed.
Is the transposed-convolution path incorrect for mixed/padded batches? or packed kernel expected to be faster?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/stage_processors.py [hunk 6]
```diff
@@ -403,6 +404,12 @@ def talker2code2wav_async_chunk(
     relative_start = emitted - base_index
     relative_end = new_end - base_index
     code_predictor_codes = torch.stack(buffer[relative_start:relative_end], dim=0).contiguous()
+    if true_finished and context_length < chunk_size:
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/stage_processors.py:407]: how much faster is it to pad? can you provide throughput/latency effect? 
if its beneficial, should we pad the startup chunks too? if we pad startup chunks - does that mean that all batches for the codec are uniform and we dont need separate implementation for uniform and non-uniform?

**Review by vklimkov-nvidia — COMMENTED** [review]
reviewed b248225.
i think the proposed change is redundant, we already know valid codec frames, there is no need to recompute (by checking values of speech codes). just pass this as meta data.

I would also want to see the effect of the proposed change. i'd like to avoid changes for the sake of changes. the implementation is supposed to work with packed sequences of variable lengths. I dont have a good idea how much faster it would be to force same shape. And if it is faster - why dont we go all the way and force input chunks also to have same shape.

**Review by vklimkov-nvidia — COMMENTED** [review]
b23c1b0b - LGTM. good, focused commit. if it was a separate PR we could have merged it straight away. I wonder though if it moves the needle in terms of performance. throughput/latency before and after the change would be good.

**Review by vklimkov-nvidia — COMMENTED** [review]
c5a29e174 - LGTM. same as with FSQ, great change, easy to review. separate performance numbers to justify would be great.

**Review by vklimkov-nvidia — COMMENTED** [review]
79250 - LGTM. great catch. easy to review, would be merged immediately if a separate PR. same as with FSQ and snake activation optimizations, performance numbers would be good to have.

File: tools/easymagpie_vllm_omni/README.md [hunk 7]
```diff
@@ -15,6 +15,11 @@ Model definition and pipeline registration live in
 [`vllm_plugin_easymagpie_omni/`](vllm_plugin_easymagpie_omni/).
 Deployment knobs are in [`deploy/easymagpie.yaml`](deploy/easymagpie.yaml).
 
+The two-stage configuration selects `EasyMagpieCodecGPUGenerationWorker` to pack
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/README.md:18]: no need to have this in README, just keep it in the comments.
if there is need for implementation documentation we can address it separately.

**Review by vklimkov-nvidia — COMMENTED** [review]
2a47b4 - we introduce custom model runner, overriding "sample_tokens". we intercept output processing and instead of running `to(cpu)` for each model output independently, we concatenate all outputs, run `to(cpu)` once and then split back.

generally i'd like to get rid of all the custom implementations in our plugin, keep only model definition and upstream all the optimizations/API changes to vllm-omni. that's why i am a bit reluctant to introduce custom runner, unless there is a strong evidence that it truly boosts performance.

I'd argue that if it makes a difference (please provide numbers) we should create PR to vllm omni with either this approach or with optional async GPU->CPU copy. We already have that for inputs and outputs i believe, just this specific output is still copied synchronously in vllm-omni.
